### PR TITLE
nixos/murmur: support setting custom stateDir, user and group

### PR DESCRIPTION
**Pull request by DCsunset**
This PR supports setting custom state directory, user and group for the Murmur service:
- The custom state directory is helpful when you want to manage the state in a different place for backup or sync.
- The user and group options allow the service to run as existing users so that permissions can be managed more easily

The `stateDir`, `user`, and `group` options are common in other modules as well. It should be easy for users to use and bring no breaking changes.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/networking/murmur.nix [hunk 1]
```diff
@@ -297,14 +315,14 @@ in
   };
 
   config = mkIf cfg.enable {
-    users.users.murmur = {
+    users.users.murmur = mkIf (cfg.user == "murmur") {
       description     = "Murmur Service user";
       home            = cfg.stateDir;
       createHome      = true;
       uid             = config.ids.uids.murmur;
       group           = "murmur";
```
Comment (felixsinger) [inline nixos/modules/services/networking/murmur.nix:323]: That needs to be adjusted as well.

**Review by felixsinger — CHANGES_REQUESTED** [review]

**Review by felixsinger — CHANGES_REQUESTED** [review]
```
<policy user="murmur">
            <allow own="net.sourceforge.mumble.murmur"/>
</policy>
```

Policy user needs to be changed as well.

File: nixos/modules/services/networking/murmur.nix [hunk 2]
```diff
@@ -343,6 +351,7 @@ in
         ProtectKernelModules = true;
         ProtectKernelTunables = true;
         ProtectSystem = "full";
+        ReadWritePaths = [ cfg.stateDir ];
```
Comment (felixsinger) [inline nixos/modules/services/networking/murmur.nix:372]: This won't have any effect at the moment, and also goes beyond what the patch says. Please remove.

File: nixos/modules/services/networking/murmur.nix [hunk 3]
```diff
@@ -69,6 +69,14 @@ in
         '';
       };
 
+      stateDir = mkOption {
+        type = types.str;
```
Comment (felixsinger) [inline nixos/modules/services/networking/murmur.nix:91]: `types.path`

**Review by felixsinger — CHANGES_REQUESTED** [review]
```
environmentFile = mkOption {
        type = types.nullOr types.path;
        default = null;
        example = "/var/lib/murmur/murmurd.env";
        ...
};
```

Make use of cfg.stateDir in the example.

Looks good otherwise.

**Comment by felixsinger** [comment]
@GrahamcOfBorg test mumble

**Comment by felixsinger** [comment]
For some reason apparmor gets triggered because of denied accesses, but I can't see why.

@thoughtpolice Since you are listed as the maintainer of the Mumble tests can you help here?

**Comment by tnias** [comment]
With the trailing slash there are no more DENIED messages in the nixosTests.mumble's log.

```diff
diff --git a/nixos/modules/services/networking/murmur.nix b/nixos/modules/services/networking/murmur.nix
index ef1dcde16748..3d8a5ca96298 100644
--- a/nixos/modules/services/networking/murmur.nix
+++ b/nixos/modules/services/networking/murmur.nix
@@ -414,7 +414,7 @@ in
         r ${config.environment.etc."lsb-release".source},
         owner rwk ${cfg.stateDir}/murmur.sqlite,
         owner rw ${cfg.stateDir}/murmur.sqlite-journal,
-        owner r ${cfg.stateDir},
+        owner r ${cfg.stateDir}/,
         r /run/murmur/murmurd.pid,
         r /run/murmur/murmurd.ini,
         r ${configFile},
```

**Comment by DCsunset** [comment]
Sorry to have overlooked the trailing slash. I've added it in this PR.

**Comment by felixsinger** [comment]
@GrahamcOfBorg test mumble

**Review by felixsinger — APPROVED** [review]

File: nixos/modules/services/networking/murmur.nix [hunk 4]
```diff
@@ -386,9 +394,9 @@ in
 
         r ${config.environment.etc."os-release".source},
         r ${config.environment.etc."lsb-release".source},
-        owner rwk /var/lib/murmur/murmur.sqlite,
-        owner rw /var/lib/murmur/murmur.sqlite-journal,
-        owner r /var/lib/murmur/,
+        owner rwk ${cfg.stateDir}/murmur.sqlite,
+        owner rw ${cfg.stateDir}/murmur.sqlite-journal,
+        owner r ${cfg.stateDir},
```
Comment (felixsinger) [inline nixos/modules/services/networking/murmur.nix]: Please add the trailing slash here as this commit is breaking it.

**Review by felixsinger — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/murmur.nix (hunk 4 above)
Comment (DCsunset) [reply nixos/modules/services/networking/murmur.nix]: I think it has been added in yesterday's commit.

Comment (tnias) [reply nixos/modules/services/networking/murmur.nix]: The PR consists of two commits. The first breaks the test (by omitting the trailing slash) and the second fixes the issue introduced by the first commit.

Comment (DCsunset) [reply nixos/modules/services/networking/murmur.nix]: Thanks for the pointer. I've rebased the commits.

**Review by felixsinger — APPROVED** [review]

**Review by tnias — APPROVED** [review]

**Pull request merged**